### DISCUSSION TITLE
Self-signed certificates

### DISCUSSION THREAD

**Discussion by tedstriker** [jellyfin/jellyfin-androidtv#3153] [General]
Would it be possible to load a self-signed CA certificate into the app or at least ignore the validation. The reason is, that Android TV doesn't allow to import root certificates.
What do you guys think?

**Comment by nielsvanvelzen** [comment]
No, we will not add such feature to the app because it is unsafe. You can add your own certificates to the Android certificate store or get a free one from services like Let's Encrypt.

**Reply by anionDev** [reply]
>because it is unsafe

No. Using your own certificate is not unsafe, especially inside your home-net which is probably the usual setup for most users. Your browser may mark it as unsafe as long as there is no root-certificate which is added to the machine's ca-certificate-store. But if you trust your self-signed certificate there is nothing wrong about using it. It is much better than falling back to unencrypted protocols like HTTP anyway.

> get a free one from services like Let's Encrypt

If your jellyfin-server is accessed through the internet then yes, this the (usual) solution.
But if your jellyfin-server is in the same local network than in your android-tv then using Let's Encrypt is more like a workaround than a solution. It should not be required. The design and purpose of PKI is that everyone is able to use his own CA-certificate in his own network. Therefore the certificate must be installed on the clients belonging to the network.

> You can add your own certificates to the Android certificate store

While I agree and endorse that adding your own CA-certificate to the machine's ca-certificate-store would be the properly solution for this issue we see: this is not a doable solution unfortunately because there is no other way to achieve this than using lowlevel-debug-tools like adb. Not a solution for a normal user.

**Reply by ChekhWasTaken** [reply]
> No, we will not add such feature to the app because it is unsafe.

unsafe is not whitelisting user certificates for the app, which forces into using http connection for android clients.
iOS provides ability to trust to the certificate on system level, making it trusted for all installed apps.

i mean lets pause and think about it for a moment: network administrators have an ability to create custom root CAs for their private networks. otherwise, that feature should not be shipped with every single major OS.
i don't see this shouldn't be respected for an app, where l belive not a small chunk of users run it in local or otherwise private networks. the fact that i'm running on private network doesn't mean i shouldn't be able to use the benefits of TLS.

**Comment by tedstriker** [comment]
As far as I know, Android TV doesn't allow users to add their own CA.
Hence my question, to allow adding the CA cert to the app instead of the system.

But if you know a way to do add it to the TV itself, I'd be more than happy to go that route.

Let's Encrypt only works, if you expose a host to the internet. This hasn't happens yet.

Also I'm curious to learn how a skipped certificate validation is more unsafe than the already implemented way to connect without any encryption at all. Honestly I never understood that idea.

**Comment by tedstriker** [comment]
@nielsvanvelzen would you mind to give me a hint? Others might benefit from it as well.

**Comment by thornbill** [comment]
You do not need a publicly accessible website to get a valid certificate from Let's Encrypt: https://letsencrypt.org/docs/challenge-types/

**Comment by tedstriker** [comment]
It still seems like I either open ports 80 (HTTP-01) or 443 (TLS-ALPN-01) for the LE-bot _or_  I need to buy a domain, use one of the supported DNS providers (DNS-01) and move the generated certificate to a directory, where Jellyfin can use it. How this is better than trusting my own, self-signed certificate is beyond my understanding.

As I lack the skills to change the app on my own, I'd be the beggar here. So I admit defeat. I asked what you guys think and you kindly let me know.

Meanwhile I just opened the unencrypted http port for my TV, so I can use the app. Entering credentials on an unencrypted channel feels wrong, but I'm not dying on that hill.

Thanks for your time and patience.

**Comment by Gylesie** [comment]
Hello @tedstriker, I have the same set-up as you and I understand the struggle. In the end, I concluded to the same workaround - having to have an unsecure connection for android tv clients. It is a pity that there is no way to force trust a certain certificate in the jellyfin client. Even more so, that there is no proper way to download a certificate into the trust store of an android tv itself. Don't get me wrong, I don't mean to criticize the devs, because most people probably wouldn't benefit from this.

**Comment by rvrosm** [comment]
use [this method](https://mobile.mrw0l05zyn.cl/android/analisis-dinamico/captura-de-trafico-de-red/burp-suite) successfully installed root ca on android tv:

```
openssl x509 -inform PEM -subject_hash_old -in root.crt | head -1
Subject hash old = 9a5ba575
adb push root.crt /sdcard/Download/

adb root
adb shell
mount -o rw,remount /
cp /sdcard/Download/root.crt /system/etc/security/cacerts
cd /system/etc/security/cacerts
mv root.crt 9a5ba575.0
chmod 644 9a5ba575.0
reboot
```

this `root.crt` is `caddyserver`'s [ACME Server](https://caddyserver.com/docs/caddyfile/directives/acme_server)'s root CA. I use `caddyserver` to serve jellyfin on [https://jellyfin.lan](https://jellyfin.lan)

**Reply by gbakeman** [reply]
I believe this requires having root?

**Comment by dgmora** [comment]
> But if you trust your self-signed certificate there is nothing wrong about using it. It is much better than falling back to unencrypted protocols like HTTP anyway.

Indeed, trying to make this "safer" and not allowing self signed certificates is making  people waste time in useless hops or falling back to http. If browsers let you use self signed certificates with a warning, I don't see why should the jellyfin app behave differently. This is making things unsafer, rather than safer.

**Reply by superboo07** [reply]
I agree, this has actually forced me to be *less* secure. I don't like having someones arbitrary restrictions dictate how unsecure my server is. There should be a box that says "allow ssl cert" ffs

**Comment by gbakeman** [comment]
This issue is especially important to me now as I no longer have a public domain (which I was only using for internal services, anyways.) I've already rolled a PKI including self-signed CA and a couple of TLS Server certs. These install just fine on every device I own except the SHIELD. As far as I can tell, there is no way to install your certificate without rooting it, and I'm reading conflicting info that this may break Widevine DRM for other services. So rooting is out of the question. Thus far I've disabled HTTPS redirecting on my proxy server, so I've directed Jellyfin to go straight to HTTP. But it would be nice to have Jellyfin whitelist/accept our self-signed cert. and connect using HTTPS. I know it's annoying putting a workaround like this into your app when the OS should be handling it, unfortunately AndroidTV and SHIELD in particular seem lacking in the area of custom PKI.

Related issues

- https://github.com/jellyfin/jellyfin-androidtv/issues/1127
- https://github.com/jellyfin/jellyfin-androidtv/issues/3385
- https://github.com/jellyfin/jellyfin-androidtv/issues/4811
- https://github.com/jellyfin/jellyfin-androidtv/issues/4231

**Comment by r0607003Dries** [comment]
I too was trying to implement HTTPS by using a self signed certificate and I got it working in the browser.
But when it came to the windows client it didn't even work after I had installed the certificate on my machine, which admittedly 
may have been an issue specific to me.

But I would have been fine with that and accepted that I would probably go through my browser to avoid using HTTP when using Jellyfin remotely...
Until I noticed that you store the password to my certificate in plain text in an unencrypted xml-file located (in my case) under:

/srv/jellyfin/config/config/network.xml 

A file of which the permissions look like this by default:

-rw-r--r-- 1 root root  998 Mar 10 18:43 network.xml

Readable by everyone, unencrypted, plain text password.
And you say self signed certificates are unsafe??

I am using Jellyfin in docker btw.

Please fix this.

**Comment by superboo07** [comment]
yeah this is bullshit, there should be a way to make self signed certificates work. I don't expose my jellyfin to the internet, I just want ssl to work on my phone. I have the certificate added in my certificate store, doesn't work. rediculous to have someones rediculous security standards emposed on me, inarguably this is signicantly less secure because I HAVE to use http because https won't work.

**Comment by maluueu** [comment]
For gods sake, just add a trust this CA  / bypass CA check button ...
I run my own CA, where every managed device (Win, Linux, iOS, iPadOS, Android) gets the CA certs into their certificate store automatically. Only Android TV devices don't support that ...

Your solution is for me to change my domain to a public one, just so that I can get a valid LE cert via DNS.
You can't be serious @nielsvanvelzen

**Reply by superboo07** [reply]
yeah its frankly unhinged the way jellyfin handles this since its expected that it will run on android tvs where adding a CA to the trust store isn't even possible on most.

**Reply by ChekhWasTaken** [reply]
i have a FireTV and i was able to push my root CA and install it there.
[Wholphin](https://github.com/damontecres/Wholphin/tree/main) was able to pick it up, although i had to allow TLS 1.2 on the nginx.

i've used following commands to install the certificate on FireTV.
```
# prerequisite: enable ADB debugging: Settings -> Device & Software -> Developer options -> ADB debugging
# turn it off again when you're done.
# confirmed to work with FireOS 7.7.1.3 and Wholphin v1.0.0
adb push <your-certificate-file> /storage/emulated/0/Download
adb shell am start -a "android.intent.action.VIEW" -d "file:///storage/emulated/0/Download/<your-certificate-file>" -t "application/x-x509-ca-cert"
```

**Reply by maluueu** [reply]
@ChekhWasTaken this only works on really old Android versions. Also AFAIK FireOS is not AndroidTV but rather plain Android 6/7/8/9.

When you try this on a newer Android TV version (e.g. 14) it errors our because Android expects the intent to be launched via the original Settings app and not ADB. Only problem is, there is no way to launch this intent from the settings app because Android TV does not have a "Add CA cert" button there.

**Reply by superboo07** [reply]
I can concur, its not a solution. even if ADB worked for this it wouldn't be a solution. shouldn't have to enter a command line just to trust my CA. very disappointing to still see developers of this project shrugging off this issue as though its a user problem, because its **not**. this is a very arbitrary restriction jellyfin has no good reason to apply

**Comment by thornbill** [comment]
I'm just going to lock this discussion. There is nothing productive coming out of this hostility on a closed three year old discussion.
